File: 5.最长回文子串.cpp
/*
 * @lc app=leetcode.cn id=5 lang=cpp
 *
 * [5] 最长回文子串
 */

// @lc code=start
class Solution {
public:
    string longestPalindrome(string s) {
        string newStr = "#";
        for (int i = 0; i < s.size(); i++) {
            newStr.push_back(s[i]);
            newStr.push_back('#');
        }
        int n = newStr.size();
        int p[n], center = 0, maxRight = 0, maxCenter = 0, maxLen = 0;
        p[0] = 0;
        for (int i = 1; i < n; i++) {
            if (i >= maxRight) {
                int right = i;
                while (right + 1 < n && 2 * i - right > 0 && newStr[right+1] == newStr[2*i-right-1]) {
                    right++;
                }
                center = i;
                maxRight = right;
                p[i] = maxRight - i;
            }
            else if (maxRight - i < p[2*center-i]) {
                p[i] = maxRight - i;
            }
            else if (maxRight - i == p[2*center-i]) {
                int right = maxRight;
                while (right + 1 < n && 2 * i - right > 0 && newStr[right+1] == newStr[2*i-right-1]) {
                    right++;
                }
                center = i;
                maxRight = right;
                p[i] = maxRight - i;
            }
            else {
                p[i] = p[2*center-i];
            }
            if (p[i] > maxLen) {
                maxLen = p[i];
                maxCenter = i;
            }
        }
        return s.substr((maxCenter - maxLen) / 2, maxLen);
    }
};
// @lc code=end

File: 5.最长回文子串.py
#
# @lc app=leetcode.cn id=5 lang=python3
#
# [5] 最长回文子串
#

# @lc code=start
class Solution:
    def longestPalindrome(self, s: str) -> str:
        s1 = "@#" + "#".join(s) + "#$"
        n = len(s1)
        dp = [0] * n
        ans = s[0]
        center, maxRight = 0, 0
        for i in range(1, n-1):
            if i < maxRight:
                dp[i] = min(maxRight-i, dp[2*center-i])
            while s1[i+dp[i]+1] == s1[i-dp[i]-1]:
                dp[i] += 1
            if i + dp[i] > maxRight:
                center, maxRight = i, i + dp[i]
            if (dp[i] > len(ans)):
                ans = s[(i-dp[i]-1)//2:(i+dp[i]-1)//2]
        return ans
# @lc code=end

